rln/src/merkle_tree.rs

```diff
@@ -1,251 +1,231 @@
+                io::ErrorKind::InvalidInput,
+                "index exceeds set size",
+            ));
+        }
+        self.nodes.insert((self.depth, index), leaf);
+        self.recalculate_from(index);
+        self.next_index = max(self.next_index, index + 1);
+        Ok(())
     }
 
-    #[must_use]
-    pub fn root(&self) -> H::Hash {
-        self.nodes[0]
+    pub fn update_next(&mut self, leaf: H::Fr) -> io::Result<()> {
+        self.set(self.next_index, leaf)?;
+        Ok(())
     }
 
-    pub fn set(&mut self, leaf: usize, hash: H::Hash) {
-        self.set_range(leaf, once(hash));
-        self.next_index = max(self.next_index, leaf + 1);
+    pub fn delete(&mut self, index: usize) -> io::Result<()> {
+        let leaf = H::default_leaf();
+        self.set(index, leaf)?;
+        Ok(())
     }
 
-    pub fn set_range<I: IntoIterator<Item = H::Hash>>(&mut self, start: usize, hashes: I) {
-        let index = self.num_leaves() + start - 1;
-        let mut count = 0;
-        // TODO: Error/panic when hashes is longer than available leafs
-        for (leaf, hash) in self.nodes[index..].iter_mut().zip(hashes) {
-            *leaf = hash;
-            count += 1;
+    pub fn proof(&self, index: usize) -> io::Result<MerkleProof<H>> {
+        if index >= self.tree_size() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "index exceeds set size",
+            ));
         }
-        if count != 0 {
-            self.update_nodes(index, index + (count - 1));
-            self.next_index = max(self.next_index, start + count);
+        let mut witness = Vec::<(H::Fr, u8)>::with_capacity(self.depth);
+        let mut i = index;
+        let mut depth = self.depth;
+        loop {
+            i ^= 1;
+            witness.push((self.get_node(depth, i), (1 - (i & 1)).try_into().unwrap()));
+            i >>= 1;
+            depth -= 1;
+            if depth == 0 {
+                break;
+            }
         }
+        assert_eq!(i, 0);
+        Ok(MerkleProof(witness))
     }
 
-    fn update_nodes(&mut self, start: usize, end: usize) {
-        debug_assert_eq!(levels(start), levels(end));
-        if let (Some(start), Some(end)) = (parent(start), parent(end)) {
-            for parent in start..=end {
-                let child = first_child(parent);
-                self.nodes[parent] = H::hash_node(&self.nodes[child], &self.nodes[child + 1]);
+    pub fn verify(&self, leaf: &H::Fr, witness: &MerkleProof<H>) -> io::Result<bool> {
+        if witness.length() != self.depth {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "witness length doesn't match tree depth",
+            ));
+        }
+        let mut acc = *leaf;
+
+        for w in witness.0.iter() {
+            if w.1 == 0 {
+                acc = H::hash(&[acc, w.0]);
+            } else {
+                acc = H::hash(&[w.0, acc]);
             }
```

Comment:
Referring to this previous comment, this if-else blocks also needs to be swapped. This did not happen to error out because the logic was swapped symmetrically in both functions i.e., `verify` and `proof`. 
```suggestion
            if w.1 == 0 {
                acc = H::hash(&[w.0, acc]);
            } else {
               acc = H::hash(&[acc, w.0]);     
            }
```